Userland/Libraries/LibSQL/TreeNode.cpp

```diff
@@ -0,0 +1,284 @@
+}
+
+u32 TreeNode::down_pointer(size_t ix) const
+{
+    VERIFY(ix < m_down_pointers.size());
+    return m_down_pointers[ix];
+}
+
+RefPtr<TreeNode> TreeNode::down_node(size_t ix) const
+{
+    VERIFY(ix < m_down_pointers.size());
+    auto ptr = m_down_pointers[ix];
+    auto ret = m_down_nodes[ix];
+    if (!ret) {
+        VERIFY(ptr != 0);
+        auto buffer = tree()->read_block(ptr);
+        size_t offset;
+        ret = adopt_ref(*new TreeNode(tree(), this, ptr, buffer, offset));
+    } else {
+        VERIFY(ret->m_pointer == ptr);
+    }
+    return ret;
+}
+
+RefPtr<TreeNode> TreeNode::node_for(Key const& key)
+{
+    dbgln("[{}] node_for(Key {})", m_pointer, (String) key);
+    dump();
+    if (is_leaf()) {
+        return this;
+    }
+    for (size_t ix = 0; ix < size(); ix++) {
+        if (key < m_entries[ix]) {
+            dbgln("[{}] {} < {} ({} -> {})",
+                m_pointer, (String) key, (String) m_entries[ix],
+                ix, m_down_nodes[ix]->pointer());
+            return down_node(ix)->node_for(key);
+        }
+    }
+    dbgln("[{}] {} >= {} ({} -> {})",
+        m_pointer, (String) key, (String) m_entries[size() - 1],
+        size(), m_down_nodes[size()]->pointer());
+    return down_node(size())->node_for(key);
+}
+
+Optional<u32> TreeNode::get(Key& key)
+{
+    dbgln("[{}] find({})",
+        m_pointer, (String) key);
+    dump();
+    for (auto ix = 0u; ix < size(); ix++) {
+        if (key < m_entries[ix]) {
+            if (is_leaf()) {
+                dbgln("[{}] {} < {} -> 0",
+                    m_pointer, (String) key, (String) m_entries[ix]);
+                return {};
+            } else {
+                dbgln("[{}] {} < {} ({} -> {})",
+                    m_pointer, (String) key, (String) m_entries[ix],
+                    ix, m_down_nodes[ix]->pointer());
+                return down_node(ix)->get(key);
+            }
+        }
+        if (key == m_entries[ix]) {
+            dbgln("[{}] {} == {} -> {}",
+                m_pointer, (String) key, (String) m_entries[ix],
+                m_entries[ix].pointer());
+            key.pointer(m_entries[ix].pointer());
+            return m_entries[ix].pointer();
+        }
+    }
+    if (is_leaf()) {
+        dbgln("[{}] {} > {} -> 0",
+            m_pointer, (String) key, (String) m_entries[size() - 1]);
+        return {};
+    } else {
+        dbgln("[{}] {} > {} ({} -> {})",
+            m_pointer, (String) key, (String) m_entries[size() - 1],
+            size(), m_down_nodes[size()]->pointer());
+        return down_node(size())->get(key);
+    }
+}
+
+void TreeNode::serialize(ByteBuffer& buffer) const
+{
+    u32 sz = size();
+    serialize_to(buffer, sz);
+    for (auto ix = 0u; ix < size(); ix++) {
+        auto& entry = m_entries[ix];
+        serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[ix]);
+        entry.serialize(buffer);
+    }
+    serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[size()]);
```

Comment:
you can just append a `u` to a number to make it unsigned, like `0u`

Reply:
Done